Add read/write specialisation for IOContext{AnnIO}

tecosaur opened the pull request:
Now that we have `AnnotatedIOBuffer` and `StyledStrings`, I decided to revisit my "tell me about this struct" utility, and make it less of a kludge.

Along the way I ran into what I think is essentially a rather rough edge that can be smoothed over with a few extra method specialisations.

See the commit message for details on this what's been done, it's changed a few times now.

vtjnash commented:
Two of those seem like what I would have expected the AbstractPipe would already do for forwarding these to the underlying implementation

tecosaur commented:
Running into this behaviour took me by surprise, but I think it's because `AnnotatedIOBuffer` implements `pipe_reader` and `pipe_writer` to point to the `IOBuffer` inside it. The generic `write` method for `IOContext` then writes to the `pipe_writer`, hence the need for this specialisation.

vtjnash commented:
Okay, I see this calls instead `StyledStrings._ansi_writer` because of overloading StyledStrings, rather than first unwrapping the IOContext. I think that suggests that it should unwrap any IOContext before dispatching to `StyledStrings._ansi_writer` for these types? Otherwise this looks like just too much of a narrow special case, that can be tricked simply by cases such as `IOContext{Any}`

tecosaur commented:
Thanks for taking a further look into this. Do I read you correctly in that you're suggesting we resolve this issue by tackling the general case where reads/writes with an IOContext behave differently from reads/writes to the wrapped IO?

vtjnash commented:
yes

tecosaur commented:
Cool, I'll have a look at doing that and see about updating this PR to take that approach instead.

tecosaur commented:
Hmmm, my first attempt at this hasn't gone well

```julia
# We want reads/writes involving an `IOContext` to dispatch to to underlying `IO`,
# rather than on `IOContext`.
write(io::IOContext, x::Any) = write(io.io, x)
read(io::IOContext, x::Any)  = read(io.io, x)
# To resolve method ambiguities
write(io::IO, x::AbstactChar) = write(io.io, x)
write(io::IO, x::AbstactString) = write(io.io, x)
write(io::IO, x::Union{String,SubString{String}}) = write(io.io, x)
write(io::IO, x::UInt8) = write(io.io, x)
```

(when I take `julia +nightly` and eval the above in a `begin ... end` block) produces

```

SYSTEM (REPL): showing an error caused an error

SYSTEM (REPL): caught exception of type Unreachable reached at 0x7f0b8cb176c6

[28873] signal 4 (2): Illegal instruction
in expression starting at none:0
print at ./strings/io.jl:250
unknown function (ip: 0x7f0b8cb176f6)
print at ./strings/io.jl:46
...

zsh: illegal hardware instruction (core dumped)  julia +nightly
18:09:05 ❯ 
```

I'm also not sure if there's a good general method do deal with all the potential method ambiguities introduced.

I'm confident about the methods for the `IOContext{AnnotatedIOBuffer}` case since we can know ahead of time which methods need to be implemented and that no ambiguities are introduced, but I'm struggling to see how to make a general solution work here.

If anybody has thoughts on this, I'd love to hear them.

Seelengrab commented:
> I'm also not sure if there's a good general method do deal with all the potential method ambiguities introduced.

Possibly related: https://github.com/JuliaLang/julia/issues/47385

tecosaur committed:
Add specialised AnnotatedString pipe read/writes

Ensure that when an AnnotatedIOBuffer is wrapped in an IOContext (or
similar AnnotatedPipe-based construct), that writes of annotated
strings/chars and reading out an AnnotatedString is unimpeded by the
IOContext wrapping.

Without these specialisations, the generic pipe_reader/pipe_writer
fallbacks will directly access the underlying IOBuffer and annotations
will be lost.

There are a number of scenarios in which one might want to combine an
AnnotatedIOBuffer and IOContext (for example setting the compact
property). Losing annotations in such scenarios is highly undesirable.
It is of particular note that this can arise in situations where you
can't unwrap the IOContext as needed, for example when passing IO to a
function you do not control (which is currently extremely hard to work
around).

Getting this right is a little difficult, and a few approaches have been
tried. Initially, I added IOContext{AnnotatedIOBuffer} specialisations
to show.jl, but arguably it's a bit of a code smell to specialise in
this way (and Jameson wasn't happy with it, with concerns that it could
be tricked by IOContext{Any}).

    # So that read/writes with `IOContext` (and any similar `AbstractPipe` wrappers)
    # work as expected.
    write(io::IOContext{AnnotatedIOBuffer}, s::Union{AnnotatedString, SubString{<:AnnotatedString}}) =
        write(io.io, s)
    write(io::AnnotatedIOBuffer, c::AnnotatedChar) = write(io.io, c)

Then I tried making it so that IOContext writes dispatched on the
wrapped IO type, but of course that broke cases like IOContext{IOBuffer}
with :color=>true.

    # So that read/writes with `IOContext` (and any similar `AbstractPipe` wrappers)
    # work as expected.
    write(io::AbstractPipe, s::Union{AnnotatedString, SubString{<:AnnotatedString}}) =
        write(pipe_writer(io), s)
    write(io::AbstractPipe, c::AnnotatedChar) = write(pipe_writer(io), c)

Finally, we have the current AbstractPipe + Annotated type
specialisation, which IOContext is just an instance of. To avoid
behaving too broadly, we need to confirm that the underlying IO is
actually an AnnotatedIOBuffer. I'm still not happy with this, only idea
I've had other than implementing IOContext{AnnotatedIOBuffer} methods
that actually seems viable, and I've had trouble soliciting help from
other people brainstorming here.

If somebody can implement something cleaner here in the future, I'd be
thrilled.

tecosaur commented:
After a rather productive chat with Nathan Zimmerberg on slack, I'm thinking that:
- It probably wouldn't be a good idea to always pass through to the underlying IO of an `IOContext`
- We can fix this headache just as effectively by making a "pass through to the `pipe_writer`/`pipe_reader`" method in `annotated.jl`

Let me know what you think!

tecosaur commented:
Update: this breaks the behaviour of `IOContext{IOBuffer}` with `:color => true`. Maybe we should just go with the `IOContext{AnnotatedIOBuffer}` specialisation after all?

tecosaur commented:
I suppose the alterative would be that we try to shoehorn the type of `pipe_writer(io)` into the dispatch with a method like so:

```julia
# So that read/writes with `IOContext` (and any similar `AbstractPipe` wrappers)
# work as expected.
function write(io::AbstractPipe, s::Union{AnnotatedString, SubString{<:AnnotatedString}})
    if pipe_writer(io) isa AnnotatedIOBuffer
        write(pipe_writer(io), s)
    else
        invoke(write, Tuple{IO, typeof(s)}, io, AbstractString)
    end
end
# Can't be part of the `Union` above because it introducec method ambiguities
function write(io::AbstractPipe, c::AnnotatedChar)
    if pipe_writer(io) isa AnnotatedIOBuffer
        write(pipe_writer(io), c)
    else
        invoke(write, Tuple{IO, typeof(c)}, io, AbstractChar)
    end
end
```

I don't think this can work for read though, but that is just a convenience method.

tecosaur commented:
If anyone has thoughts on this, I'd love to hear them. Currently, I find the alternative mentioned in my last comment a bit ugly, but if there's no other input I'll just go with that.

tecosaur commented:
I've updated this PR. @vtjnash please let me know what you think of the current state.

tecosaur commented:
Let me know if I'm being too ambitious, but I'm going to speculatively tag this with `backport-1.11` since it (IMO) improves the state of this feature in the 1.11 release.

tecosaur commented:
I've just resolved a recently-introduced merge conflict.

More broadly, I still think this code looks a bit ugly, but it seems to work, and it's functionality that I think is rather worth having.

Should we go ahead and merge this, and let any improvements/nicer implementations come along in the future?

fingolfin reviewed (approved):
Just had a chat with @tecosaur about this over on Slack, and he pointed out the commit message here which is excellent and provides further details.

All in all, I share the mild unease about the code in here, and I think there may be need to change some things in the IO system, but this PR is not the place for it. And if this code causes any problems down the road, well, the only way to find out is to try it in the wild?

With that I'd be happy to have this merged at any time.

tecosaur commented:
(just pushed a typo fix, no changes to the code)
